File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/Sudoku/Sudoku_VisualBasic/SolverGenerator/Techniques/NakedSingleTechnique.vb
'--------------------------------------------------------------------------
' 
'  Copyright (c) Microsoft Corporation.  All rights reserved. 
' 
'  File: NakedSingleTechnique.vb
'
'  Description: Implements the Sudoku naked single technique.
' 
'--------------------------------------------------------------------------

Imports Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Collections

Namespace Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Techniques
	''' <summary>Implements the naked single elimination technique.</summary>
	<Serializable>
	Public NotInheritable Class NakedSingleTechnique
		Inherits EliminationTechnique
		''' <summary>Initialize the technique.</summary>
		Public Sub New()
		End Sub

		''' <summary>Gets the difficulty level of this technique.</summary>
		Friend Overrides ReadOnly Property DifficultyLevel() As UInteger
			Get
				Return 2
			End Get
		End Property

		Public Overrides Overloads Function Equals(ByVal obj As Object) As Boolean
			Return TypeOf obj Is NakedSingleTechnique
		End Function

		Public Overrides Function GetHashCode() As Integer
			Return Me.GetType().GetHashCode()
		End Function

		''' <summary>Runs this elimination technique over the supplied puzzle state and previously computed possible numbers.</summary>
		''' <param name="state">The puzzle state.</param>
		''' <param name="possibleNumbers">The previously computed possible numbers.</param>
		''' <param name="numberOfChanges">The number of changes made by this elimination technique.</param>
		''' <returns>Whether more changes may be possible based on changes made during this execution.</returns>
        Friend Overrides Function Execute(ByVal state As PuzzleState, ByVal exitEarlyWhenSoleFound As Boolean,
                    ByVal possibleNumbers()() As FastBitArray, <System.Runtime.InteropServices.Out()> ByRef numberOfChanges As Integer,
                    <System.Runtime.InteropServices.Out()> ByRef exitedEarly As Boolean) As Boolean
            numberOfChanges = 0
            exitedEarly = False

            ' Eliminate impossible numbers based on numbers already set in the grid
            For i = 0 To state.GridSize - 1
                For j = 0 To state.GridSize - 1
                    ' If this cell has a value, we use it to eliminate numbers in other cells
                    If state(i, j).HasValue Then
                        Dim valueToEliminate = state(i, j).Value

                        ' eliminate numbers in same row
                        For y = 0 To state.GridSize - 1
                            If possibleNumbers(i)(y)(valueToEliminate) Then
                                numberOfChanges += 1
                                possibleNumbers(i)(y)(valueToEliminate) = False
                            End If
                        Next y

                        ' eliminate numbers in same column
                        For x = 0 To state.GridSize - 1
                            If possibleNumbers(x)(j)(valueToEliminate) Then
                                numberOfChanges += 1
                                possibleNumbers(x)(j)(valueToEliminate) = False
                            End If
                        Next x

                        ' eliminate numbers in same box
                        Dim boxStartX = (i \ state.BoxSize) * state.BoxSize
                        For x = boxStartX To boxStartX + state.BoxSize - 1
                            Dim boxStartY = (j \ state.BoxSize) * state.BoxSize
                            For y = boxStartY To boxStartY + state.BoxSize - 1
                                If possibleNumbers(x)(y)(valueToEliminate) Then
                                    numberOfChanges += 1
                                    possibleNumbers(x)(y)(valueToEliminate) = False
                                End If
                            Next y
                        Next x
                    End If
                Next j
            Next i

            Return False
        End Function
	End Class
End Namespace

File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/Sudoku/Sudoku_CSharp/SolverGenerator/Techniques/NakedSingleTechnique.cs
//--------------------------------------------------------------------------
// 
//  Copyright (c) Microsoft Corporation.  All rights reserved. 
// 
//  File: NakedSingleTechnique.cs
//
//  Description: Implements the Sudoku naked single technique.
// 
//--------------------------------------------------------------------------

using System;
using Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Collections;

namespace Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Techniques
{
	/// <summary>Implements the naked single elimination technique.</summary>
	[Serializable]
	public sealed class NakedSingleTechnique : EliminationTechnique
	{
		/// <summary>Initialize the technique.</summary>
		public NakedSingleTechnique(){}

		/// <summary>Gets the difficulty level of this technique.</summary>
		internal override uint DifficultyLevel { get { return 2; } }

        public override bool Equals(object obj)
        {
            return obj is NakedSingleTechnique;
        }

        public override int GetHashCode()
        {
            return GetType().GetHashCode();
        }

		/// <summary>Runs this elimination technique over the supplied puzzle state and previously computed possible numbers.</summary>
		/// <param name="state">The puzzle state.</param>
		/// <param name="possibleNumbers">The previously computed possible numbers.</param>
		/// <param name="numberOfChanges">The number of changes made by this elimination technique.</param>
		/// <returns>Whether more changes may be possible based on changes made during this execution.</returns>
		internal override bool Execute(
			PuzzleState state, bool exitEarlyWhenSoleFound, 
			FastBitArray[][] possibleNumbers, out int numberOfChanges, out bool exitedEarly)
		{
			numberOfChanges = 0;
			exitedEarly = false;

			// Eliminate impossible numbers based on numbers already set in the grid
			for (int i = 0; i < state.GridSize; i++)
			{
				for (int j = 0; j < state.GridSize; j++)
				{
					// If this cell has a value, we use it to eliminate numbers in other cells
					if (state[i, j].HasValue)
					{
						byte valueToEliminate = state[i, j].Value;

						// eliminate numbers in same row
						for (int y = 0; y < state.GridSize; y++)
						{
							if (possibleNumbers[i][y][valueToEliminate])
							{
								numberOfChanges++;
								possibleNumbers[i][y][valueToEliminate] = false;
							}
						}

						// eliminate numbers in same column
						for (int x = 0; x < state.GridSize; x++)
						{
							if (possibleNumbers[x][j][valueToEliminate])
							{
								numberOfChanges++;
								possibleNumbers[x][j][valueToEliminate] = false;
							}
						}

						// eliminate numbers in same box
						int boxStartX = (i / state.BoxSize) * state.BoxSize;
						for (int x = boxStartX; x < boxStartX + state.BoxSize; x++)
						{
							int boxStartY = (j / state.BoxSize) * state.BoxSize;
							for (int y = boxStartY; y < boxStartY + state.BoxSize; y++)
							{
								if (possibleNumbers[x][y][valueToEliminate])
								{
									numberOfChanges++;
									possibleNumbers[x][y][valueToEliminate] = false;
								}
							}
						}
					}
				}
			}

			return false;
		}
	}
}